src/segment-loader.js

```diff
@@ -2044,34 +2063,138 @@ export default class SegmentLoader extends videojs.EventTarget {
+        message: 'Quota exceeded error with append of a single segment of content',
+        // To prevent any possible repeated downloads for content we can't actually
+        // append, blacklist forever.
+        blacklistDuration: Infinity
+      });
+      this.trigger('error');
+      return;
+    }
+
+    // To try to resolve the quota exceeded error, clear back buffer and retry. This means
+    // that the segment-loader should block on future events until this one is handled, so
+    // that it doesn't keep moving onto further segments. Adding the call to the call
+    // queue will prevent further appends until waitingOnRemove_ and
+    // quotaExceededErrorRetryTimeout_ are cleared.
+    //
+    // Note that this will only block the current loader. In the case of demuxed content,
+    // the other load may keep filling as fast as possible. In practice, this should be
+    // OK, as it is a rare case when either audio has a high enough bitrate to fill up a
+    // source buffer, or video fills without enough room for audio to append (and without
+    // the availability of clearing out seconds of back buffer to make room for audio).
+    // But it might still be good to handle this case in the future as a TODO.
+    this.waitingOnRemove_ = true;
+    this.callQueue_.push(this.appendToSourceBuffer_.bind(this, {segmentInfo, type, bytes}));
+
+    const currentTime = this.currentTime_();
+    // Try to remove as much audio and video as possible to make room for new content
+    // before retrying.
+    const timeToRemoveUntil = currentTime - MIN_BACK_BUFFER;
+
+    this.logger_(`On QUOTA_EXCEEDED_ERR, removing video from 0 to ${timeToRemoveUntil}`);
+    this.sourceUpdater_.removeVideo(0, timeToRemoveUntil, () => {
+      this.logger_(`On QUOTA_EXCEEDED_ERR, removing audio from 0 to ${timeToRemoveUntil}`);
+      this.sourceUpdater_.removeAudio(0, timeToRemoveUntil, () => {
+        this.logger_(`On QUOTA_EXCEEDED_ERR, retrying append in ${MIN_BACK_BUFFER}s`);
+        this.waitingOnRemove_ = false;
+        // wait the length of time alotted in the back buffer to prevent wasted
+        // attempts (since we can't clear less than the minimum)
+        this.quotaExceededErrorRetryTimeout_ = window.setTimeout(() => {
+          this.logger_('On QUOTA_EXCEEDED_ERR, re-processing call queue');
+          this.quotaExceededErrorRetryTimeout_ = null;
+          this.processCallQueue_();
+        }, MIN_BACK_BUFFER * 1000);
+      });
     });
+  }
+
+  handleAppendError_({segmentInfo, type, bytes}, error) {
+    // if there's no error, nothing to do
+    if (!error) {
+      return;
+    }
+
+    if (error.code === QUOTA_EXCEEDED_ERR) {
+      this.handleQuotaExceededError_({segmentInfo, type, bytes});
+      // A quota exceeded error should be recoverable with a future re-append, so no need
+      // to trigger an append error.
+      return;
+    }
+
+    this.error(`${type} append of ${bytes.length}b failed for segment ` +
```

Comment:
should we log the error itself too? Maybe the message or callstack in it could have more info for us?

Reply:
Yeah, probably a good idea to add that in.